src/workers/helpers.js

```diff
@@ -1,53 +1,238 @@
+      num_genes: ngenes,
+      genes: gene_info,
+      annotations: state[step].fetchCellAnnotations().columnNames(),
+    };
+
+    return output;
+  } else if (step === "rna_quality_control") {
+    let output = {};
+    var blocks = state["inputs"].fetchBlockLevels();
+    if (blocks === null) {
+      blocks = ["default"];
+      output.data = { default: state[step].fetchMetrics() };
+    } else {
+      let metrics = state[step].fetchMetrics({ copy: "view" });
+      let bids = state["inputs"].fetchBlock();
+      output.data = splitMetricsByBlock(metrics, blocks, bids);
+    }
+
+    let listed = state[step].fetchFilters();
+    output.thresholds = splitThresholdsByBlock(listed, blocks);
+
+    return output;
+  } else if (step === "adt_quality_control") {
+    var output = {};
+
+    var blocks = state["inputs"].fetchBlockLevels();
+    if (blocks === null) {
+      blocks = ["default"];
+      output.data = { default: state[step].fetchMetrics() };
+    } else {
+      let metrics = state[step].fetchMetrics({ copy: "view" });
+      let bids = state["inputs"].fetchBlock();
+      output.data = splitMetricsByBlock(metrics, blocks, bids);
+    }
+
+    let listed = state[step].fetchFilters();
+    output.thresholds = splitThresholdsByBlock(listed, blocks);
+
+    // We don't use sums for filtering but we do report it in the metrics,
+    // so we just add some NaNs to the thresholds for consistency.
+    for (const [k, v] of Object.entries(output.thresholds)) {
+      v.sums = NaN;
+    }
+
+    return output;
+  } else if (step === "crispr_quality_control") {
+    let output = {};
+    var blocks = state["inputs"].fetchBlockLevels();
+    if (blocks === null) {
+      blocks = ["default"];
+      output.data = { default: state[step].fetchMetrics() };
+    } else {
+      let metrics = state[step].fetchMetrics({ copy: "view" });
+      let bids = state["inputs"].fetchBlock();
+      output.data = splitMetricsByBlock(metrics, blocks, bids);
+    }
+
+    let listed = state[step].fetchFilters();
+    output.thresholds = splitThresholdsByBlock(listed, blocks);
+
+    return output;
+  } else if (step === "cell_filtering") {
+    let remaining = 0,
+      discard_vec = null;
+    const discardBuff = state[step].fetchDiscards();
+    if (discardBuff) {
+      discardBuff.forEach((x) => {
+        remaining += x == 0;
+      });
+      discard_vec = discardBuff.slice();
+    } else {
+      let available = state["inputs"].hasAvailable();
+      remaining = state["inputs"]
+        .fetchCountMatrix(available[0])
+        .numberOfColumns();
+    }
+    let output = { retained: remaining, discard: discard_vec };
+    return output;
+  } else if (step === "rna_normalization") {
+    return {};
+  } else if (step === "adt_normalization") {
+    return {};
+  } else if (step === "crispr_normalization") {
+    return {};
+  } else if (step === "feature_selection") {
+    let output = {
```

Comment:
`copy` needs to be `true` here as these are Float64Arrays, which will have their buffers transferred out. But they should be `true` by default anyway, so you can just remove the options altogether.